src/vmm/src/gdb/arch/aarch64.rs

```diff
@@ -1,62 +1,296 @@
+const PTE_ADDRESS_MASK: u64 = !0b111u64;
+
+/// Read a u64 value from a guest memory address
+fn read_address(vmm: &Vmm, address: u64) -> Result<u64, GdbTargetError> {
+    let mut buf = [0; 8];
+    vmm.guest_memory().read(&mut buf, GuestAddress(address))?;
+
+    Ok(u64::from_le_bytes(buf))
+}
+
+/// Translates a virtual address according to the Vcpu's current address translation mode.
+///
+/// To simplify the implementation we've made some assumptions about the paging setup
+/// here we just assert firstly paging is setup and these assumptions are correct
+pub fn translate_gva(vcpu_fd: &VcpuFd, gva: u64, vmm: &Vmm) -> Result<u64, GdbTargetError> {
+    // Check this virtual address is in kernel space
+    if extract_bits_64!(gva, 55, 1) == 0 {
+        return Err(GdbTargetError::GvaTranslateError);
+    }
+
+    // Translation control register
+    let tcr_el1: u64 = get_sys_reg(TCR_EL1, vcpu_fd)?;
+
+    // If this is 0 then translation is not yet ready
+    if extract_bits_64!(tcr_el1, 16, 6) == 0 {
+        return Ok(gva);
+    }
+
+    // Check 4KB pages are being used
+    if extract_bits_64!(tcr_el1, 30, 2) != 2 {
+        return Err(GdbTargetError::GvaTranslateError);
+    }
+
+    // ID_AA64MMFR0_EL1 provides information about the implemented memory model and memory
+    // management, check this is a physical address size we support
+    let pa_size = match get_sys_reg(ID_AA64MMFR0_EL1, vcpu_fd)? & 0b1111 {
+        0 => 32,
+        1 => 36,
+        2 => 40,
+        3 => 42,
+        4 => 44,
+        5 => 48,
+        _ => return Err(GdbTargetError::GvaTranslateError),
+    };
+
+    // A mask of the physical address size for a virtual address
+    let pa_address_mask: u64 = !0u64 >> (64 - pa_size);
+    // A mask used to take the bottom 12 bits of a value this is as we have a grainsize of 9
+    // asserted with our 4kb page, plus the offset of 3
+    let lower_mask: u64 = 0xFFF;
+    // A mask for a physical address mask with the lower 12 bits cleared
+    let desc_mask: u64 = pa_address_mask & !lower_mask;
+
+    let page_indicies = [
+        (gva >> 36) & lower_mask,
+        (gva >> 27) & lower_mask,
+        (gva >> 18) & lower_mask,
+        (gva >> 9) & lower_mask,
+    ];
+
+    // Transition table base register used for initial table lookup
+    // take the bottom 48 bits from the register value
+    let mut address: u64 = get_sys_reg(TTBR1_EL1, vcpu_fd)? & pa_address_mask;
+    let mut level = 0;
+
+    while level < 4 {
+        // Clear the bottom 3 bits from this address
+        let pte = read_address(vmm, (address + page_indicies[level]) & PTE_ADDRESS_MASK)?;
+        address = pte & desc_mask;
+
+        // If this is a valid table entry and we aren't at the end of the page tables
+        // then loop again and check next level
+        if (pte & 2 != 0) && (level < 3) {
+            level += 1;
+            continue;
+        }
+        break;
+    }
+
+    let page_size = 1u64 << ((9 * (4 - level)) + 3);
```

Comment:
Can you expand on the magic of this formula a bit?

Reply:
Added a comment explaining this now